File: usaco/contest/2017/open/Silver/p2_cownomics.py
task_name = "cownomics"
INF = 100000000


def get_list_from_line(dtype=int):
    return [dtype(x) for x in fin.readline().strip().split()]


def get_strs_from_line():
    return fin.readline().strip().split()



fin = open(f'{task_name}.in', 'r')
fout = open(f'{task_name}.out', 'w')

N, M = get_list_from_line()
spot = []
plain = []
for _ in range(N):
    spot.append(get_strs_from_line()[0])
for _ in range(N):
    plain.append(get_strs_from_line()[0])


ans = 0

for k in range(M):
    for p in range(k + 1, M):
        for j in range(p + 1, M):
            spot_set = set()
            plain_set = set()
            for i in range(N):
                spot_set.add((spot[i][j], spot[i][p], spot[i][k]))
                plain_set.add((plain[i][j], plain[i][p], plain[i][k]))
            if len(spot_set & plain_set) == 0:
                ans += 1
fout.write(f"{ans}\n")
fout.close()        

File: usaco/contest/2017/open/Silver/p2_cownomics.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <set>
#include <unordered_set>
using namespace std;

const int INF = 100000000;

vector<string> get_strs_from_line(istream &fin) {
    string line;
    getline(fin, line);
    vector<string> result;
    size_t pos = 0, next_pos;
    while ((next_pos = line.find(' ', pos)) != string::npos) {
        result.push_back(line.substr(pos, next_pos - pos));
        pos = next_pos + 1;
    }
    result.push_back(line.substr(pos));
    return result;
}

int main() {
    ifstream fin("cownomics.in");
    ofstream fout("cownomics.out");

    int N, M;
    fin >> N >> M;
    fin.ignore(); // Ignore the newline character after reading M

    vector<string> spot(N), plain(N);
    for (int i = 0; i < N; ++i) {
      fin >> spot[i]; //get_strs_from_line(fin)[0];
    }
    for (int i = 0; i < N; ++i) {
      fin >> plain[i]; // = get_strs_from_line(fin)[0];
    }

    int ans = 0;

    for (int k = 0; k < M; ++k) {
        for (int p = k + 1; p < M; ++p) {
            for (int j = p + 1; j < M; ++j) {
                unordered_set<string> spot_set, plain_set;
                for (int i = 0; i < N; ++i) {
                  string s_spot = string(1, spot[i][j]) + string(1, spot[i][p]) + string(1, spot[i][k]);
                  string s_plain = string(1, plain[i][j]) + string(1, plain[i][p]) + string(1, plain[i][k]);
                  spot_set.insert(s_spot);
                    plain_set.insert(s_plain);
                }
                int found = 0;
                if (spot_set.size() < plain_set.size()){
                    for (string x: spot_set){
                      if (plain_set.find(x) != plain_set.end()){
                        found = 1;
                    }
                 }
                }
                else {
                    for (string x: plain_set){
                      if (spot_set.find(x) != spot_set.end()){
                        found = 1;
                    }
                 }
                }

                    if (found == 0) ++ans;
            }
        }
    }

    fout << ans << endl;

    return 0;
}
